[ky.kiske] 코드 리뷰 요청합니다!

SimJung opened the pull request:
안녕하세요! 카이입니다. 지하철 노선도 미션을 진행하며 부족함을 많이 느낍니다...
일단 미션 1, 2는 수행 완료 했습니다만, 미션 3을 진행하며 로직들을 변경하는 와중 리뷰 요청을 드렸습니다!
현재 진행된 부분은 노선 생성, 노선 제거를 DB 연동 완료 하였으며, 노선 수정, 역 목록을 포함하는 조회 등은 아직 미완성인 상태입니다.
구현에 초점을 두다보니 리팩토링을 나중에 하려고 코드를 짰는데, 덕분에 미션 2에서 쓰였던 코드인 saveSection, deleteSection 등의 메소드는 depth가 2를 넘어가거나, 메소드의 라인이 길어지는 등, 여러 문제가 있는 코드입니다 ㅜㅜ
알고 싶은 부분은 구간(section) 정보를 DB에 저장하여 라인에서 어떤 식으로 써야하는지에 대한 로직입니다. 어떻게 저장해야 효율적으로 각 노선에서의 순서를 지킬 수 있는지 생각 하기가 어려운 것 같습니다... 미완성인 코드이지만 오늘까지 리뷰 요청을 드리기위해 부득이하게 이렇게 리뷰 요청 드립니다.
감사합니다!!

SimJung committed:
feat : StationController 기능 구현

 - showStations() 메소드를 통해 조회가 가능하도록 구현
 - deleteStation() 메소드를 통해 삭제가 가능하도록 구현

SimJung committed:
feat : 노선 생성 기능 구현

 - Line 클래스 생성
 - LineDao 클래스 생성
 - LineController 클래스 생성
 - 노선 생성 기능 구현 및 테스트 완료

SimJung committed:
feat : 노선 삭제 기능 구현

 - 노선 삭제 기능 구현 및 테스트 완료

SimJung committed:
feat : 노선 조회 기능 구현

 - 노선 전체 목록 조회 기능 구현 및 테스트 완료
 - 특정 노선 조회 기능 구현 및 테스트 완료

SimJung committed:
feat : 노선 수정 기능 구현

 - 노선 수정 기능 구현 및 테스트 완료

kakao-dino-bin committed:
modify: 요구사항 변경에 맞춰 노선 생성 수정

노선 생성시 종점 ID와 거리를 입력받도록 수정

-

kakao-dino-bin committed:
chore: 기타 변경 사항 및 백업 커밋

-

kakao-dino-bin committed:
feat: 구간 추가 기능 구현

노선 추가 후 구간이 추가되며 반환 기능 구현
생성 테스트 수행

-

kakao-dino-bin committed:
feat: 삭제 기능 추가

노선의 역 삭제 기능 추가
테스트 코드의 예외 상황 추가

-

SimJung committed:
feat : DB에 데이터 저장되도록 수정

 - JDBC 의존성 추가
 - Bean 객체 관리 위한 어노테이션 추가
 - Station 정보 DB에서 관리하도록 변경

SimJung committed:
feat : DB 조회하여 역 목록 받아오기

 - 기존의 리스트로 관리하던 역 목록을 DB를 통해 조회할 수 있도록 변경

SimJung committed:
feat : DB 조회하여 역 하나 받아오기

 - DB를 통해 역 하나의 정보를 받아올 수 있도록 변경

SimJung committed:
feat : DB로부터 역 제거하도록 변경

SimJung committed:
feat : StationService 추가

 - interface 및 class 구현하여 컨트롤러에서 서비스 이용하도록 변경

SimJung committed:
feat : 노선 DB 저장 기능 구현

 - Dao에서 관리하던 노선을 DB를 통해 저장할 수 있도록 변경

SimJung committed:
feat : 노선 DB 삭제 기능 구현

SimJung committed:
feat : LineService, SectionController 추가

 - LineController 에서 service를 사용하도록 변경
 - Section 관리를 SectionController에서 하도록 변경

SimJung committed:
refactor : 패키지 분리 및 리팩토링

 - Section 패키지 분리
 - Section 서비스 인터페이스 및 클래스 추가
 - 기타 리팩토링

kakao-dino-bin committed:
modify: line 조회 jdbc연결

이전에 라인 조회는 그저 list에서 가져오는 방식이었지만
db를 연결해서 가져오도록 변경

-

kakao-dino-bin committed:
modify: line 수정 jdbc적용

-

kakao-dino-bin committed:
modify: 구간 추가, 제거, 조회, 라인 일부 수정

구간 추가시 db에 저장되도록 구현
구간 제거, 조회시 db에서 불러와 각 상행 하행이 연결되도록 수정
라인이 시점, 종점을 갖고있도록 수정

-

kakao-dino-bin committed:
fix: 구간 추가, 제거 버그 수정

구간을 추가하거나 제거할때 제대로 안되는 버그를 수정
기능적으로 문제는 발견되지 않음

-

SimJung committed:
modify: section save & delete 로직 수정

 - section 리스트를 관리하는 일급 컬렉션 추가
 - sectinos를 이용하여 관리하도록 하고 기존의 idx를 이용한 삽입과 삭제를 수정
 - 중복 구간 추가 이슈와 구간 추가 시 거리 이슈를 fix함.

SimJung committed:
refactor: Section save, delete 과정 리팩토링

 - sectionServiceImpl의 save, delete 과정을 리팩토링
 - private 메소드를 늘려 구분함
 - Sections unit test 추가

SimJung commented:
안녕하세요! 카이입니다.
지하철 미션 3단계까지 진행을 하여 리뷰 요청 드립니다.
미션 수행이 늦어져 죄송합니다 ㅜㅜ...
SectionServiceImpl의 delete 부분을 좀 더 리팩토링 하고 싶은데 엮인 if문과 반복되는 메소드를 분리하기가 어려웠습니다.
해당 부분에 대한 방법과 여러 부족한 부분에 대한 따끔한 피드백을 받고 싶습니다. 감사합니다!!

humbledude commented on Read:
```diff
@@ -0,0 +1,13 @@
+structure{
```
임시파일인가요?
임시 파일은 커밋에 추가되지 않도록 .gitignore 를 사용하셔도 좋겠네요

humbledude commented on src/main/java/subway/section/Sections.java:
```diff
@@ -0,0 +1,59 @@
+            }
+        }
+
+        return -1L;
```
-1 이 결과없음을 뜻하기에는 좀 애매한거 같습니다
상수로 정의 또는 null 또는 exception 을 던지는 등의 명확한 액션이 좋을꺼 같습니다

humbledude commented on src/main/java/subway/section/Sections.java:
```diff
@@ -0,0 +1,59 @@
+                .orElse(null);
+    }
+
+    public Long findStationExist(Section section) {
```
find station 인데 인자로 section 이 와서 좀 헷갈리는 네이밍 같아요~

humbledude commented on src/main/java/subway/line/Line.java:
```diff
@@ -0,0 +1,69 @@
+    private Long upStationId;
+    private Long downStationId;
+
+    private List<Section> sections = new ArrayList<>();
```
sections 일급 콜렉션이 있는데 안쓰신 이유가 있을까요?

humbledude commented on src/main/java/subway/station/StationService.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.util.List;
+
+public interface StationService {
```
서비스 interface 를 만드신 이유가 있을까요?

humbledude commented on src/main/java/subway/station/StationController.java:
```diff
@@ -6,25 +6,43 @@
+    public ResponseEntity deleteStation(@PathVariable Long stationId) {
+        if (!stationService.deleteById(stationId))
+            return ResponseEntity.badRequest().build();
+        return ResponseEntity.ok().build();
```
delete 성공에 대한 응답 코드를 수정하셨는데요,
아래 기준에 맞게 잘 수정된것인지 한번 생각해 보셔도 좋겠어요

https://developer.mozilla.org/ko/docs/Web/HTTP/Methods/DELETE

humbledude commented on src/main/java/subway/section/SectionRequest.java:
```diff
@@ -1,14 +1,14 @@
     private Long upStationId;
     private Long downStationId;
-    private int distance;
+    private Integer distance;
```
int 와 Integer 를 썼을때의 차이가 뭐가 있을까요?

humbledude commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,88 @@
+        if (newLine == null) {
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
+        }
+        return ResponseEntity.created(URI.create("/lines/" + newLine.getId())).body(new LineResponse(newLine.getId(), newLine.getName(), newLine.getColor()));
```
한줄을 너무 길게 작성하지 않는게 좋겠어요 ㅎㅎ
가독성을 위해서 적절히 분리 해주세요~

humbledude commented on src/main/java/subway/section/SectionServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+        Long dest = line.getDownStationId();
+        List<Section> orderedSections = new LinkedList();
+
+        while (!cur.equals(dest)) {
```
물론 데이터가 잘 구성되면 좋겠지만..
사람일이 항상 그렇지가 않죠 ㅎㅎ
섹션이 제대로 처음부터 끝까지 잘 일렬로 구성되지 않으면 여기 무한루프에 빠지겠어요

humbledude commented on src/main/java/subway/section/SectionServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+        return checkAndAddSection(section);
+    }
+
+    private boolean checkAndAddSection(Section section) {
```
check 하고 add 한다는 두가지 작업을 하나로 하면 어떨까요?
당연히 add 하기 전에는 check 하는거고, 
add 를 하는 와중에 check 가 실패하면 예외를 발생하는 등의 로직을 하면 단순해질꺼 같습니다.

humbledude commented on src/main/java/subway/section/SectionDao.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public Section save(Section section) {
+        SimpleJdbcInsert simpleJdbcInsert = new SimpleJdbcInsert(jdbcTemplate).withTableName("section")
```
이거 insert 전용 아닌가요? update 도 잘 되나요? (제가 안써봐서 ㅎㅎ)

humbledude commented on src/main/java/subway/section/SectionServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+        if (nextSection != null && prevSection != null) {
+            deleteSectionById(nextSection.getSectionId());
+            deleteSectionById(prevSection.getSectionId());
+            save(new Section(prevSection.getUpStationId(), nextSection.getDownStationId(), prevSection.getDistance() + nextSection.getDistance(), lineId));
```
아 이 save 가 insert 전용이어서 delete 를 수반하는가 보네요~
update 로 구현하면 더 단순해 지려나..
한번 고민해 보시면 좋겠어요~

humbledude reviewed (changes requested):
안녕하세요~ 리뷰를 맡은 조슈아 입니다

이번 미션에서는 다양한 요구조건이 있는 웹앱을 다루고있는데요,
로또에서 배우셨던대로, 비즈니스 로직은 domain class 들에 구현되도록 수정해보시길 바랍니다.
지금은 사실상 주요 로직은 SectionServiceImpl 에 다 되어있네요

그렇게 구현이 되면, 유닛 테스트는 도메인 객체 중심으로 작성할 수 있게 되어 좀 더 테스트가 명확해 집니다.
(지금은 테스트가 controller 랑 service 랑 domain 이 다 엮여 있죠)

좀 어려울수 있지만 한번 도전해 보세요~

SimJung committed:
refactor: sections 클래스 상수 추가 및 메소드 리네임

 - NOT_EXIST (-1) 값을 상수로 추가
 - findStationExist 메소드 뒤에 직관성을 위해 BySection을 추가함

SimJung committed:
refactor: 메소드 호출 시 메소드명 변경

 - 수정한 findStationExistBySection을 호출하도록 변경
 - Line 클래스에서 사용되지 않는 sections를 제거함.

SimJung committed:
modify: Service 인터페이스 삭제

 - 사실상 쓸 이유가 없는 Service 인터페이스들을 삭제
 - 각 Impl 클래스들의 이름을 Service로 변경

SimJung committed:
modify: 데이터 삭제 시 반환하는 응답 코드 변경

 - 기존의 200 (ok) 코드에서 204 (noContent) 응답 코드를 반환하도록 수정

SimJung committed:
style: 개행 추가

 - 코드의 가독성을 위해 개행 추가

SimJung committed:
fix: 무한 루프 체크 로직 추가

 - 섹션이 제대로 구성되어 있지 않을 경우 RuntimeExeption이 발생하도록 수정

SimJung committed:
refactor: checkAndAdd 메소드 분리

 - check 과정을 메소드로 분리
 - 이외의 과정은 기존의 saveSection 메소드로 이동시킴

SimJung committed:
refactor: Service의 로직을 domain으로 일부 이동

 - Line 클래스에서 섹션 정보를 바탕으로 상행, 하행 종점인지 반환하도록 메소드 이동
 - SectionDao 에서 담당하던 중복 섹션 판단을 Sections에서 해주도록 변경
 - SectionService에서 담당하던 정렬, 다음 혹은 이전 섹션 찾기 등을 sections로 이동

SimJung committed:
test: 섹션 정렬 시 무한 루프 테스트 추가

 - 무한 루프를 피하기 위한 코드 테스트 및 성공

SimJung commented:
남겨주신 코멘트들을 참고하여 업데이트 해보았습니다만, 역시 service의 기능들을 한 번에 domain으로 옮기려니 막히는 부분이 많았습니다.
최대한 옮긴건가 싶다가도 다시 코드를 보면 service의 크기는 그대로인것 같아 아쉬웠습니다...
현재 단계에서 가장 큰 문제는 save와 delete 메소드에서 dao를 많이 참조하는 부분인 것 같습니다. 이를 도메인으로 옮기기에는 적절하지 않다고 판단했습니다만, 이를 해결하기 위해선 dao에서 db를 가지고 수정하는 로직을 도메인에서 담당해야 하는건지, 아니면 어떤 식으로 접근해야 할지 궁금합니다. 감사합니다!!

humbledude commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -30,12 +32,17 @@ private Sections getOrderedSection(Sections sections, Line line) {
+        Set<Long> visit = new HashSet<>();
 
-        while (!cur.equals(dest)) {
+        while (!cur.equals(dest) && visit.add(cur)) {
```
중복 체크로 탈출 이군요 👍 
근데 데이터가 중복이 아니고 중간에 끊어져 있다면??? ㅎㅎㅎ
다양한 이슈 상황이 있으니 체크 해보시길 바랍니다

SimJung replied:
헉.. 그렇군요... 새삼 다양한 상황에서의 이슈 처리의 중요성을 느끼게 되었습니다. 감사합니다!

humbledude reviewed (approved):
안녕하세요~

dao 를 참조하는 부분이 어려운 부분이 맞습니다 ㅎㅎㅎ
dao 만 없었어도 domain 객체만 남을수 있겠죠?

그말인 즉~ dao 가 없다고 코딩을 하신담에 전후에 dao 로 db 작업을 할수도 있겠죠.
예를 들어 sections 에 새로운 section 이 들어온다면 dao 로 필요한 데이터를 싹 불러와서,
dao 의존없이 sections 와 section 으로만 계산을 하고, validation 해서 이상 없다면 변경점을 dao 로 싹 저장해주면 되는거겠죠

이런 부분을 편하게 생각할 수 있는 도구로 JPA 가 있으니 한번 참고해 보세요
(편하게 쓰다가 고려할게 더 많아질수는 있지만 ㅎㅎ)

이번 미션은 여기서 마무리 하고, 다음 미션에서 관련해서 한번 domain 에 비즈니스 로직을 넣는 연습을 더 해보시면 좋겠네요~

수고 많으셨습니다.

SimJung commented:
말씀해주신 domain을 풍부하게 만드는 부분을 다음 미션에서 좀 더 생각해보도록 하겠습니다. 리뷰 해주셔서 감사합니다!!
